File: benchmarks/go/compiler_emit.go
package main

import (
	"fmt"
	"strings"
)

func emitStatement(builder *strings.Builder, functionID int, blockID int, statementID int) {
	target := (statementID + functionID + blockID) % 17
	left := (statementID*3 + functionID) % 23
	right := (blockID*5 + statementID) % 29
	fmt.Fprintf(builder, "  v%d = v%d + %d;\n", target, left, right)
}

func emitBlock(builder *strings.Builder, functionID int, blockID int, statements int) {
	fmt.Fprintf(builder, "block_%d:\n", blockID)
	for statementID := 0; statementID < statements; statementID++ {
		emitStatement(builder, functionID, blockID, statementID)
	}
	if blockID%3 == 0 {
		fmt.Fprintf(builder, "  goto block_%d;\n", blockID+1)
	} else {
		fmt.Fprintf(builder, "  acc += v%d;\n", blockID%17)
	}
}

func emitFunction(builder *strings.Builder, functionID int, blocks int, statements int) {
	fmt.Fprintf(builder, "static long f%d(long seed) {\n", functionID)
	builder.WriteString("  long acc = seed;\n")
	for i := 0; i < 24; i++ {
		fmt.Fprintf(builder, "  long v%d = seed + %d;\n", i, functionID+i)
	}
	for blockID := 0; blockID < blocks; blockID++ {
		emitBlock(builder, functionID, blockID, statements)
	}
	builder.WriteString("  return acc;\n}\n")
}

func emitProgram(functions int, blocks int, statements int) string {
	var builder strings.Builder
	builder.Grow(functions * blocks * statements * 24)
	builder.WriteString("/* generated benchmark program */\n")
	for functionID := 0; functionID < functions; functionID++ {
		emitFunction(&builder, functionID, blocks, statements)
	}
	return builder.String()
}

func checksumText(text string) int {
	checksum := 0
	for i := 0; i < len(text); i++ {
		checksum = checksum*33 + int(text[i])
	}
	return checksum
}

func main() {
	text := emitProgram(120, 10, 8)
	fmt.Printf("compiler_emit checksum: %d\n", checksumText(text))
	fmt.Printf("compiler_emit bytes: %d\n", len(text))
}

File: benchmarks/ocaml/compiler_emit.ml
(* Compiler-shaped code emission benchmark.
   Models generated C emission from a simple IR using OCaml Buffer, which is
   the realistic baseline for compiler output construction. *)

let emit_statement buf function_id block_id statement_id =
  let target = (statement_id + function_id + block_id) mod 17 in
  let left = ((statement_id * 3) + function_id) mod 23 in
  let right = ((block_id * 5) + statement_id) mod 29 in
  Buffer.add_string buf "  v";
  Buffer.add_string buf (string_of_int target);
  Buffer.add_string buf " = v";
  Buffer.add_string buf (string_of_int left);
  Buffer.add_string buf " + ";
  Buffer.add_string buf (string_of_int right);
  Buffer.add_string buf ";\n"

let emit_block buf function_id block_id statements =
  Buffer.add_string buf "block_";
  Buffer.add_string buf (string_of_int block_id);
  Buffer.add_string buf ":\n";
  for statement_id = 0 to statements - 1 do
    emit_statement buf function_id block_id statement_id
  done;
  if block_id mod 3 = 0 then (
    Buffer.add_string buf "  goto block_";
    Buffer.add_string buf (string_of_int (block_id + 1));
    Buffer.add_string buf ";\n")
  else (
    Buffer.add_string buf "  acc += v";
    Buffer.add_string buf (string_of_int (block_id mod 17));
    Buffer.add_string buf ";\n")

let emit_function buf function_id blocks statements =
  Buffer.add_string buf "static long f";
  Buffer.add_string buf (string_of_int function_id);
  Buffer.add_string buf "(long seed) {\n";
  Buffer.add_string buf "  long acc = seed;\n";
  for i = 0 to 23 do
    Buffer.add_string buf "  long v";
    Buffer.add_string buf (string_of_int i);
    Buffer.add_string buf " = seed + ";
    Buffer.add_string buf (string_of_int (function_id + i));
    Buffer.add_string buf ";\n"
  done;
  for block_id = 0 to blocks - 1 do
    emit_block buf function_id block_id statements
  done;
  Buffer.add_string buf "  return acc;\n}\n"

let emit_program functions blocks statements =
  let buf = Buffer.create (functions * blocks * statements * 24) in
  Buffer.add_string buf "/* generated benchmark program */\n";
  for function_id = 0 to functions - 1 do
    emit_function buf function_id blocks statements
  done;
  Buffer.contents buf

let checksum_text text =
  let checksum = ref 0 in
  for i = 0 to String.length text - 1 do
    checksum := (!checksum * 33) + Char.code text.[i]
  done;
  !checksum

let () =
  let text = emit_program 120 10 8 in
  Printf.printf "compiler_emit checksum: %d\n" (checksum_text text);
  Printf.printf "compiler_emit bytes: %d\n" (String.length text)
